### PR TITLE
clearify network-plugin-dir for different network plugins

### PR DESCRIPTION
**Pull request by freehan**
ref: https://github.com/kubernetes/kubernetes/issues/28563

### CODE REVIEW

File: docs/admin/network-plugins.md [hunk 1]
```diff
@@ -43,3 +43,10 @@ The plugin requires a few things:
 * Kubelet must be run with the `--network-plugin=kubenet` argument to enable the plugin
 * Kubelet must also be run with the `--reconcile-cidr` argument to ensure the IP subnet assigned to the node by configuration or the controller-manager is propagated to the plugin
 * The node must be assigned an IP subnet through either the `--pod-cidr` kubelet command-line option or the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` controller-manager command-line options.
+
+## Usage Summary
+
+* `--network-plugin=exec`: Must specify the `--network-plugin-dir` flag to the location of executables.
+* `--network-plugin=cni`: Specify the `--network-plugin-dir` flag to the location of CNI config file. Default to `/etc/cni/net.d`. Meanwhile, CNI binaries must be present in `/opt/cni/bin`.
+* `--network-plugin=kubenet`: Specify the `--network-plugin-dir` to path where the CNI `bridge` and `host-local` plugins are placed in addition to default path `/opt/cni/bin`.
```
Comment (bprashanth) [inline docs/admin/network-plugins.md]: do we need both bridge and host-local plugins? if so please emphasize this. What happens if there's only one (what breaks)? 

Also --network-plugin-dir here specifies the location of plugins, but with --network-plugin=cni --network-plugin-dir has the location of CNI files? (according to your previous section)

File: docs/admin/network-plugins.md [hunk 2]
```diff
@@ -43,3 +43,10 @@ The plugin requires a few things:
 * Kubelet must be run with the `--network-plugin=kubenet` argument to enable the plugin
 * Kubelet must also be run with the `--reconcile-cidr` argument to ensure the IP subnet assigned to the node by configuration or the controller-manager is propagated to the plugin
 * The node must be assigned an IP subnet through either the `--pod-cidr` kubelet command-line option or the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` controller-manager command-line options.
+
+## Usage Summary
+
+* `--network-plugin=exec`: Must specify the `--network-plugin-dir` flag to the location of executables.
+* `--network-plugin=cni`: Specify the `--network-plugin-dir` flag to the location of CNI config file. Default to `/etc/cni/net.d`. Meanwhile, CNI binaries must be present in `/opt/cni/bin`.
```
Comment (bprashanth) [inline docs/admin/network-plugins.md]: Something like: specifies that we use the cni plugin with cin plugin configuration located in `network-plugin-dir` and actual plugin binaries located in ..., config location defaults to ..., binary location defaults to

File: docs/admin/network-plugins.md [hunk 3]
```diff
@@ -43,3 +43,10 @@ The plugin requires a few things:
 * Kubelet must be run with the `--network-plugin=kubenet` argument to enable the plugin
 * Kubelet must also be run with the `--reconcile-cidr` argument to ensure the IP subnet assigned to the node by configuration or the controller-manager is propagated to the plugin
 * The node must be assigned an IP subnet through either the `--pod-cidr` kubelet command-line option or the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` controller-manager command-line options.
+
+## Usage Summary
+
+* `--network-plugin=exec`: Must specify the `--network-plugin-dir` flag to the location of executables.
```
Comment (bprashanth) [inline docs/admin/network-plugins.md]: How about something like: specifies that we use the exec plugin, with executables located in `--network-plugin-dir`. If `network-plugin-dir` is unspecified it defaults to ....

File: docs/admin/network-plugins.md (hunk 1 above)
Comment (freehan) [inline docs/admin/network-plugins.md]: Need both plugins. It has been emphasized above. kubenet will return failure without doing any actual operation if one of them is not present.

**Comment by freehan** [comment]
PTAL

**Comment by bprashanth** [comment]
oh, why isn't this in the release-1.3 barnch? (I don't remember where all your previous docs went either)

File: docs/admin/network-plugins.md [hunk 4]
```diff
@@ -43,3 +43,9 @@ The plugin requires a few things:
 * Kubelet must be run with the `--network-plugin=kubenet` argument to enable the plugin
 * Kubelet must also be run with the `--reconcile-cidr` argument to ensure the IP subnet assigned to the node by configuration or the controller-manager is propagated to the plugin
 * The node must be assigned an IP subnet through either the `--pod-cidr` kubelet command-line option or the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` controller-manager command-line options.
+
+## Usage Summary
+
+* `--network-plugin=exec` specifies that we use the `exec` plugin, with executables located in `--network-plugin-dir`.
+* `--network-plugin=cni` specifies that we use the `cni` network plugin with actual CNI plugin binaries located in `/opt/cni/bin` and CNI plugin configuration located in `network-plugin-dir`, config location defaults to `/etc/cni/net.d`.
```
Comment (bprashanth) [inline docs/admin/network-plugins.md]: just clarifying: 
--network-plugin=cni --network-plugin-dir=(must point to location of cni conf)
But:
--network-plugin=exec --network-plugin-dir=(must point to location of executables) 
? 
this is confusing

File: docs/admin/network-plugins.md [hunk 5]
```diff
@@ -43,3 +43,9 @@ The plugin requires a few things:
 * Kubelet must be run with the `--network-plugin=kubenet` argument to enable the plugin
 * Kubelet must also be run with the `--reconcile-cidr` argument to ensure the IP subnet assigned to the node by configuration or the controller-manager is propagated to the plugin
 * The node must be assigned an IP subnet through either the `--pod-cidr` kubelet command-line option or the `--allocate-node-cidrs=true --cluster-cidr=<cidr>` controller-manager command-line options.
+
+## Usage Summary
+
+* `--network-plugin=exec` specifies that we use the `exec` plugin, with executables located in `--network-plugin-dir`.
```
Comment (bprashanth) [inline docs/admin/network-plugins.md]: do we get an error if network-plugin-dir is empty?

File: docs/admin/network-plugins.md (hunk 4 above)
Comment (freehan) [inline docs/admin/network-plugins.md]: yeah. I know. It has been there since the initial commit...

File: docs/admin/network-plugins.md (hunk 5 above)
Comment (freehan) [inline docs/admin/network-plugins.md]: nope. It will continue till init and fail at that point if no executables can be found

**Comment by dcbw** [comment]
Explanations are accurate according to my understanding of things, though obviously it's still confusing.

LGTM

**Pull request merged**